### ISSUE TITLE
Migrating from RNCryptor-objc to RNCryptor-swift

### ISSUE THREAD

**Issue by abush12345** [RNCryptor/RNCryptor#309]
I am working on an application that is using the older objc version of RNCryptor. 

The current objc release of RNCryptor is able to decrypt the data that has been encrypted.

When I attempt to switch to the SPM Swift and decrypt previously encrypted data using that, I get a UnknownHeader error.

Is there anything I need to do to ensure that the newer swift version can decrypt data that was encrypted using the older objc version?

It was originally encrypted like this:

    NSData *data = [value dataUsingEncoding:NSUTF8StringEncoding];
    NSData *encryptedData = [RNEncryptor encryptData:data
                                        withSettings:kRNCryptorAES256Settings
                                            password:encryptionKey
                                               error:&error];

return [encryptedData base64EncodedStringWithOptions:0];

and previously decrypted like this:

    NSData *encryptedData = [[NSData alloc] initWithBase64EncodedString:value options:0];
    NSData *decryptedData = [RNDecryptor decryptData:encryptedData
                                        withPassword:encryptionKey
                                               error:&error];
        
        return [[NSString alloc] initWithData:decryptedData encoding:NSUTF8StringEncoding];

Swift decryption which is failing with unknown header:

            let data: Data = Data(base64Encoded: encrypted)! 
            let decryptedData = try RNCryptor.decrypt(data: data, withPassword: password)
            let decryptedString = String(data: decryptedData, encoding: .utf8) 

What am I missing?  Thanks for any help!

**Comment by rnapier** [comment]
What version of the ObjC library is being used here? What is first several bytes of ‘data’?I expect you’re either using a very very old version of the ObjC library (which I can help you with), or you’re incorrectly decoding your data in Swift earlier in the process. On Feb 6, 2024, at 8:04 PM, abush12345 ***@***.***> wrote:﻿
I am working on an application that is using the older objc version of RNCryptor.
The current objc release of RNCryptor is able to decrypt the data that has been encrypted.
When I attempt to switch to the SPM Swift and decrypt previously encrypted data using that, I get a UnknownHeader error.
Is there anything I need to do to ensure that the newer swift version can decrypt data that was encrypted using the older objc version?
It was originally encrypted like this:
NSData *data = [value dataUsingEncoding:NSUTF8StringEncoding];
NSData *encryptedData = [RNEncryptor encryptData:data
                                    withSettings:kRNCryptorAES256Settings
                                        password:encryptionKey
                                           error:&error];

return [encryptedData base64EncodedStringWithOptions:0];
and previously decrypted like this:
NSData *encryptedData = [[NSData alloc] initWithBase64EncodedString:value options:0];
NSData *decryptedData = [RNDecryptor decryptData:encryptedData
                                    withPassword:encryptionKey
                                           error:&error];
    
    return [[NSString alloc] initWithData:decryptedData encoding:NSUTF8StringEncoding];

Swift decryption which is failing with unknown header:
        let data: Data = Data(base64Encoded: encrypted)! 
        let decryptedData = try RNCryptor.decrypt(data: data, withPassword: password)
        let decryptedString = String(data: decryptedData, encoding: .utf8) 

What am I missing?  Thanks for any help!

**Comment by abush12345** [comment]
Thanks for your reply Rob.

"What version of the ObjC library is being used here? What is first several bytes of ‘data’?I expect you’re either using a very very old version of the ObjC library (which I can help you with),"

It is very, very old.   The original developers took the approach of copying the base file set into the codebase and its been ignored ever since.  

It works with your current version of the objc version, and Ive confirmed that the Swift decoder is being passed exactly the same data as we are sending to the objc version.

We have a lot of existing users, so If we have to do a slow migration, that is ok, but if there is something you could suggest to speed that up and allow an immediate update, I would love to hear it.

Thanks again for your help.

Trying to find something that will help you guess which version its from, there is nothing explicit about version that Ive found.
In RNCryptor.m:
NSString *const kRNCryptorErrorDomain = @"net.robnapier.RNCryptManager";
const uint8_t kRNCryptorFileVersion = 2;

In RNDecryptor.m:
static const NSUInteger kPreambleSize = 2;

- (BOOL)updateOptionsForPreamble:(NSData *)preamble
{
  const uint8_t *bytes = [preamble bytes];

  // See http://robnapier.net/blog/rncryptor-hmac-vulnerability-827 for information on the v1 bad HMAC
#ifdef RNCRYPTOR_ALLOW_V1_BAD_HMAC
  if (bytes[0] == 1) {
    self.options = bytes[1];
    self.hasV1HMAC = YES;
    return YES;
  }
#endif

  if (bytes[0] == kRNCryptorFileVersion) {
    self.options = bytes[1];
    return YES;
  }

  return NO;
}

**Comment by rnapier** [comment]
Yeah, that's what I expected. This is v2 of the format. It was replaced in 2013 and was never ported to Swift.

The only difference between v2 and v3 is how multi-byte passwords are handled. v2 had [a significant bug that truncated them](https://github.com/RNCryptor/RNCryptor/issues/77). If you know that your passwords are always in ASCII, then you can just adjust the code to accept 2 as well as 3. (An easier solution may be to just fix the leading byte from 0x02 to 0x03 before passing it to the decryptor.)

If you need to deal with arbitrary passwords, then the fix will be similar to https://github.com/RNCryptor/RNCryptor/commit/2c3cae0e677c1aa4d841b655c82bcb0d4086bd60.

Depending on your needs, I can discuss making the changes for you, but my open source time is a bit limited right now. I can definitely work in a small consulting job, but it may take a while before I made this change. I would accept a PR, however, to support v2 decryption in Swift.

**Comment by abush12345** [comment]
I will double check, but Im pretty confident that in the context it is being used, the passwords are provably all ASCII, so I will experiment with changing the leading byte and see how that plays out.

Thanks for your support Rob, and for the great software, its worked very well for us for years (too long, arguably :)

I will poke around a little.
